src/Evolution/Systems/CurvedScalarWave/Worldtube/PunctureField.cpp

```diff
@@ -0,0 +1,441 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "Evolution/Systems/CurvedScalarWave/Worldtube/PunctureField.hpp"
+
+#include "DataStructures/DataBox/Prefixes.hpp"
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/DynamicBuffer.hpp"
+#include "DataStructures/Tags/TempTensor.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "DataStructures/Variables.hpp"
+#include "Evolution/Systems/CurvedScalarWave/Tags.hpp"
+#include "NumericalAlgorithms/LinearOperators/PartialDerivatives.hpp"
+#include "Utilities/Gsl.hpp"
+
+namespace CurvedScalarWave::Worldtube {
+
+void compute_puncture_field(
+    gsl::not_null<Variables<tmpl::list<
+        Tags::Psi, ::Tags::dt<Tags::Psi>,
+        ::Tags::deriv<Tags::Psi, tmpl::size_t<3>, Frame::Inertial>>>*>
+        evolved_vars,
+    const tnsr::I<DataVector, 3, Frame::Inertial>& coords, const double time,
+    const double orbital_radius, const double bh_mass, const size_t order) {
+  if (order == 0) {
+    compute_puncture_field_0(evolved_vars, coords, time, orbital_radius,
+                             bh_mass);
+  } else {
+    ERROR(
+        "Puncture order is only implemented up to order 0 but you requested "
+        "order "
+        << order);
+  }
+}
+
+void compute_puncture_field_0(
+    gsl::not_null<Variables<tmpl::list<
+        Tags::Psi, ::Tags::dt<Tags::Psi>,
+        ::Tags::deriv<Tags::Psi, tmpl::size_t<3>, Frame::Inertial>>>*>
+        evolved_vars,
+    const tnsr::I<DataVector, 3, Frame::Inertial>& coords, const double time,
+    const double orbital_radius, const double BH_mass) {
+  const size_t grid_size = get<0>(coords).size();
+  evolved_vars->initialize(grid_size);
+  const double r0 = orbital_radius;
+  const double M = BH_mass;
+  const double w = 1. / (r0 * sqrt(r0));
+  const double t = time;
+
+  const double charge_pos_x = r0 * cos(w * time);
+  const double charge_pos_y = r0 * sin(w * time);
+  const double charge_pos_z = 0.;
+
+  const DataVector& x = get<0>(coords);
+  const DataVector& y = get<1>(coords);
+  const DataVector& z = get<2>(coords);
+
+  const DataVector Dx = x - charge_pos_x;
+  const DataVector Dy = y - charge_pos_y;
+  const DataVector Dz = z - charge_pos_z;
+
+  // we use a dynamic buffer even though the size is known at compile
+  // time because TempBuffer only accepts 256 arguments and takes much
+  // longer to compile. The performance loss was measured to be about 10%.
+  DynamicBuffer<DataVector> temps(66, grid_size);
+
+  const double d_0 = 1.0 / r0;
+  const double d_1 = 3 * M;
```

Comment:
is it easy to regex-replace [space][number][space with [space][number].0[space]? If not, I wouldn't worry about it... This is auto-generated code anyway

Reply:
it's not that easy unfortunately because the spaces only appear after I `clang-format`. Could probably done with a bit of work but might not be worth the effort - should be the same binary after all.

Reply:
nevermind, fixed it now